File: cdp-agentkit-core/python/cdp_agentkit_core/actions/address_reputation.py
import re
from collections.abc import Callable

from cdp import Address
from pydantic import BaseModel, Field, field_validator

from cdp_agentkit_core.actions import CdpAction

ADDRESS_REPUTATION_PROMPT = """
This tool checks the reputation of an address on a given network. It takes:

- network: The network the address is on (e.g. "base-mainnet")
- address: The Ethereum address to check

Important notes:
- This tool will not work on base-sepolia, you can default to using base-mainnet instead
- The wallet's default address and its network may be used if not provided
"""


class AddressReputationInput(BaseModel):
    """Input argument schema for checking address reputation."""

    address: str = Field(..., description="The Ethereum address to check")
    network: str = Field(..., description="The network to check the address on")

    @field_validator("address")
    @classmethod
    def validate_address(cls, v: str) -> str:
        """Validate that the provided address is a valid Ethereum address.

        Args:
            v (str): The address string to validate

        Returns:
            str: The validated address string

        Raises:
            ValueError: If the address format is invalid

        """
        if not re.match(r"^0x[a-fA-F0-9]{40}$", v):
            raise ValueError("Invalid Ethereum address format")
        return v


def check_address_reputation(address: str, network: str) -> str:
    """Check the reputation of an address.

    Args:
        address (str): The Ethereum address to check
        network (str): The network the address is on

    Returns:
        str: A string containing the reputation json data or error message

    """
    try:
        address = Address(network, address)
        reputation = address.reputation()
        return str(reputation)
    except Exception as e:
        return f"Error checking address reputation: {e!s}"


class AddressReputationAction(CdpAction):
    """Address reputation check action."""

    name: str = "address_reputation"
    description: str = ADDRESS_REPUTATION_PROMPT
    args_schema: type[BaseModel] | None = AddressReputationInput
    func: Callable[..., str] = check_address_reputation

File: cdp-agentkit-core/typescript/src/actions/cdp/address_reputation.ts
import { Wallet, Address } from "@coinbase/coinbase-sdk";
import { z } from "zod";

import { CdpAction } from "./cdp_action";

const ADDRESS_REPUTATION_PROMPT = `
This tool checks the reputation of an address on a given network. It takes:

- network: The network to check the address on (e.g. "base-mainnet")
- address: The Ethereum address to check

Important notes:
- This tool will not work on base-sepolia, you can default to using base-mainnet instead
- The wallet's default address and its network may be used if not provided
`;

/**
 * Input schema for address reputation check.
 */
export const AddressReputationInput = z
  .object({
    address: z
      .string()
      .regex(/^0x[a-fA-F0-9]{40}$/, "Invalid Ethereum address format")
      .describe("The Ethereum address to check"),
    network: z.string().describe("The network to check the address on"),
  })
  .strip()
  .describe("Input schema for address reputation check");

/**
 * Check the reputation of an address.
 *
 * @param wallet - The wallet instance
 * @param args - The input arguments for the action
 * @returns A string containing reputation data or error message
 */
export async function checkAddressReputation(
  args: z.infer<typeof AddressReputationInput>,
): Promise<string> {
  try {
    const address = new Address(args.network, args.address);
    const reputation = await address.reputation();
    return reputation.toString();
  } catch (error) {
    return `Error checking address reputation: ${error}`;
  }
}

/**
 * Address reputation check action.
 */
export class AddressReputationAction implements CdpAction<typeof AddressReputationInput> {
  public name = "address_reputation";
  public description = ADDRESS_REPUTATION_PROMPT;
  public argsSchema = AddressReputationInput;
  public func = checkAddressReputation;
}
